niri-config/src/lib.rs

```diff
@@ -944,6 +945,145 @@ where
+            if let Some(unwanted_child) = children.next() {
+                ctx.emit_error(DecodeError::unexpected(
+                    unwanted_child,
+                    "node",
+                    "expected no more than one child",
+                ));
+            }
+            PresetWidth::decode_node(child, ctx).map(Some).map(Self)
+        } else {
+            Ok(Self(None))
+        }
+    }
+}
+
+impl<S> knuffel::Decode<S> for Binds
+where
+    S: knuffel::traits::ErrorSpan,
+{
+    fn decode_node(
+        node: &knuffel::ast::SpannedNode<S>,
+        ctx: &mut knuffel::decode::Context<S>,
+    ) -> Result<Self, DecodeError<S>> {
+        expect_only_children(node, ctx);
+
+        let mut seen_keys = HashSet::new();
+
+        let mut binds = Vec::new();
+
+        for child in node.children() {
+            match Bind::decode_node(child, ctx) {
+                Err(e) => {
+                    ctx.emit_error(e);
+                }
+                Ok(bind) => {
+                    if seen_keys.insert(bind.key) {
+                        binds.push(bind);
+                    } else {
+                        // TODO: add context of the source span for the original bind?
+                        // i couldn't figure out how to easily do this
+                        ctx.emit_error(DecodeError::unexpected(
+                            &child.node_name,
+                            "keybind",
+                            "duplicate keybind",
+                        ));
+                    }
+                }
+            }
+        }
+
+        Ok(Self(binds))
+    }
+}
+
+impl<S> knuffel::Decode<S> for Bind
+where
+    S: knuffel::traits::ErrorSpan,
+{
+    fn decode_node(
+        node: &knuffel::ast::SpannedNode<S>,
+        ctx: &mut knuffel::decode::Context<S>,
+    ) -> Result<Self, DecodeError<S>> {
+        expect_only_children(node, ctx);
+
+        let key = node
+            .node_name
+            .parse::<Key>()
+            .map_err(|e| DecodeError::conversion(&node.node_name, e.wrap_err("invalid keybind")))?;
+
+        let mut children = node.children();
+
+        if let Some(child) = children.next() {
+            for unwanted_child in children {
+                ctx.emit_error(DecodeError::unexpected(
+                    unwanted_child,
+                    "node",
+                    "only one action is allowed per keybind",
+                ));
+            }
+            let action = Action::decode_node(child, ctx)?;
+            Ok(Self { key, action })
+        } else {
+            ctx.emit_error(DecodeError::missing(
+                node,
+                "expected an action for this keybind",
+            ));
+            // This value isn't used. Decoding will not succeed, because we just emitted an error.
+            // Even if it somehow were used, spawn without args is a no-op.
+            Ok(Self {
+                key,
+                action: Action::Spawn(vec![]),
+            })
```

Comment:
Why not just return Err(DecodeError) with the error right above?

Reply:
I don't know how i thought to write a comment, but did not think to write in the comment why i did it this way. This is done like so, such that `Binds` will have a happy path on return, adding dummy data to the actual binds, but crucially remembering that the `key` was parsed correctly. In other words, this ensures that "duplicate keybind" is emitted even when there is an empty block. I did miss the case where there is a block with an invalid action, though.

Reply:
i think returning Err is not fine, because then we also suppress the  "duplicate keybind". my latest commit documents why we do this a lot better, though. i agree, the previous comment was unhelpful at best and did not at all communicate the intent